### PR TITLE
Issue 290 Part 2

### LINKED ISSUE #290: String representation of objects
It would be nice if we could add the `__repr__` method to biotite objects such as sequences, etc. Representations are especially useful when working in Jupiter notebook.

### PR DESCRIPTION
**Pull request by MaxGreil**
Fixes #290

New __repr__ method for classes:

- CodonTable
- Alignment
- SubstitutionMatrix
- Location
- Feature
- Annotation
- AnnotatedSequence
- Atom

Classes AtomArray and AtomArrayStack can use __repr__ method from respective superclass

### Questions

@padix-key I think for classes CodonTable and SubstitutionMatrix are new __eq__ and __ne__ methods necessary? What do you think?

### CODE REVIEW

**Comment by padix-key** [comment]
Thank you for you effort! :+1:

>  I think for classes CodonTable and SubstitutionMatrix are new eq and ne methods necessary? What do you think?

I agree. Could you add them in this PR?

> Classes AtomArray and AtomArrayStack can use repr method from respective superclass

Could you explain? The class `_AtomArrayBase` has no implemented `__repr__()` method.

For the `CodonTable` class I would like to request to put the `starts` parameter name into the repr-string. I think this makes it clearer for the user what the second parameter means.

**Comment by MaxGreil** [comment]
> Could you explain? The class _AtomArrayBase has no implemented __repr__() method.

You are right @padix-key , I will also implement the `__repr__` method for classes `AtomArray` and `AtomArrayStack`. Regarding these two classes, should the __repr__` method look like in the discussion for issue #290

`array([
    Atom(array([1.0, 2.0, 3.0]), chain_id="A", res_id=1, ...),
    ...
])
`
?

**Comment by padix-key** [comment]
I would say so. However, there is a name conflict, as the outer `array()` refers to `structure.array()` and the inner `array()` refers to `numpy.array()`. Maybe we change the inner function to `'np.array([...])'`? For `AtomArrayStack` I would propose something like
```
stack([
    array(...),
    array(...)
])
```

**Comment by MaxGreil** [comment]
@padix-key I applied the changes

File: src/biotite/structure/atoms.py [hunk 1]
```diff
@@ -660,7 +673,17 @@ def __init__(self, length):
...
+
+    def __repr__(self):
+        """Represent AtomArray as a string for debugging."""
+        atoms = ''
+        for i in range(0, self.array_length()):
+            if len(atoms) == 0:
+                atoms = self.get_atom(i).__repr__()
+            else:
+                atoms = atoms + ',\n' + self.get_atom(i).__repr__()
+        return f'array([{atoms}])'
```
Comment (padix-key) [inline src/biotite/structure/atoms.py:685]: Could you put each atom on a separate line with indentation? Currently, the output is
```
array([Atom(...)
...
Atom(...)])
```

but I think
```
array([
    Atom(...)
    ...
    Atom(...)
])
```
would be better readable.

File: src/biotite/structure/atoms.py [hunk 2]
```diff
@@ -909,7 +932,17 @@ def __init__(self, depth, length):
...
+
+    def __repr__(self):
+        """Represent AtomArray as a string for debugging."""
+        arrays = ''
+        for i in range(0, self.stack_depth()):
+            if len(arrays) == 0:
+                arrays = self.get_array(i).__repr__()
+            else:
+                arrays = arrays + ',\n' + self.get_array(i).__repr__()
+        return f'stack([{arrays}])'
```
Comment (padix-key) [inline src/biotite/structure/atoms.py:944]: Similar to the comment above, i would like to request another layer of indentation and an empty line between each model to visibly separate them.

**Review by padix-key — CHANGES_REQUESTED** [review]
I have a few minor points, which I highlighted in the comments below.

**Review by padix-key — APPROVED** [review]
Thank you. I will merge the PR now. The failing AppVeyor tests are an independent issue in the package setup.

**Pull request merged**